### PR TITLE
Adds type annotations to base.py

### PR DESCRIPTION
**Pull request by happz**
And many other files, since base.py is one of the core pieces of tmt,
and many other modules depend on it or are used by base.py code. This
means mypy is allowed to look around, and reports also conflicts between
existing annotations and new ones that must be amended.

### CODE REVIEW

**Comment by happz** [comment]
Follow up for #1178 - big, ugly, adding type annotations to `base.py`.

**Comment by happz** [comment]
@thrix how's your work on `StepData` going? I see plenty of mypy complains related to `StepData`-like `data` objects having wrong keys, now when mypy is allowed to look into `base.py`, and I'm thinking `StepData` need to be resolved soon to make annotations click. Can I help with that patch in any way?

**Comment by psss** [comment]
Thanks for preparing the changes, looking into it. For the record, currently 13 tests are failing.

File: tests/test/export/test.sh
Comment (happz) [inline tests/test/export/test.sh:22]: Note to myself: why did the order change? export still begins with `dict(name=self.name)`, `name` should have been the first item logged. With Python 3.7, keys should be presented in the inserted order, and I did not change things that much...

Comment (psss) [reply tests/test/export/test.sh:22]: We still support `el8` with `python 3.6` so... we probably should not rely on the order.

Comment (happz) [reply tests/test/export/test.sh:22]: Things are fine. The test checks fmf id, and I did indeed change the order of fields. In any case, it's an export, and YAML on top of that, `grep` and presuming any particular order of fields will be always unstable. Dropping `-` should be perfectly safe.

**Review by psss — CHANGES_REQUESTED** [review]
Thanks much for looking into this! Added a bunch of comments. Still reading through `base.py`, will finish later today or tomorrow.

File: tmt/steps/discover/fmf.py
Comment (happz) [reply tmt/steps/discover/fmf.py:299]: Yes, but if `path` is `None`, `fmf_root = self.step.plan.my_run.tree.root` happens, and `root` is `Optional[str]`.

If we are sure `self.step.plan.my_run.tree.root` cannot/should not be `None` when we take the `else` branch, then I can add `assert` there and drop `Optional` trait.

File: tmt/beakerlib.py
Comment (happz) [reply tmt/beakerlib.py:345]: `dependency` is of type `Require`, which is a `Union[str, RequireFmfId]`. This is perfectly compatible with `require`-like containers, but not with `recommend`-like ones because fmf id is not allowed under `recommend` key. Since the content of `require` and `recommend` keys is squashed into a single iterable, a check is needed to make sure we're comparing only the `str` subtype against `require` content.

File: tmt/cli.py
Comment (happz) [reply tmt/cli.py:715]: That means `tmt.base.ExportFormat(format_)` instead of `.DICT`?

File: tmt/cli.py
Comment (happz) [reply tmt/cli.py:1177]: Actually, there are not enough `cast()` calls, as I just realized:

```python
def plans_show(context: click.core.Context, **kwargs: Any) -> None:
    ...
    tmt.Plan._save_context(context)
    for plan in context.obj.tree.plans():
        plan.show()
        echo()
```

`context.obj` is annotated as `Any`, mypy cannot infer it's actually `tmt.cli.ContextObject`. This means, wherever `context.obj` is accessed, we should cast it to `ContextObject` so mypy could catch up and follow the trail:

```python
def plans_show(context: click.core.Context, **kwargs: Any) -> None:
    ...
    tmt.Plan._save_context(context)
    for plan in cast(ContextObject, context.obj).tree.plans():
        plan.show()
        echo()
```

So, I should review `cli.py` and add these calls. I suppose that would make you raise your eyebrows even more :) Another idea: how about adding a custom subclass of `click.core.Context` which would only serve as a type stub, e.g.:

```python
class Context(click.core.Context):
  obj: ContextObject  # yeah, maybe we should also rename this class, there's way too many contexts and objects...

...

# Then this should work:
def plans_show(context: Context, **kwargs: Any) -> None:
    ...
    assert context.obj is not None
    tmt.Plan._save_context(context)
    for plan in context.obj.tree.plans():
        plan.show()
        echo()
```

No more `cast()`, but we should replace all `click.Context` annotations in code with the custom class, to make things right (all commands that take `context` parameter, for example...). This plan would actually be my first option, sounds like the least expensive introduction of the "context.obj is of `ContextObject` type" concept. verbose, absolutely, but also fairly correct.

What do you think?

File: tmt/steps/discover/fmf.py
Comment (happz) [reply tmt/steps/discover/fmf.py:320]: I believe that's a valid observation and concern. I believe there is a solution, although hard to pick one.

`Plan.my_run` is `Optional[Run]`, because `Plan.__init__()` accepts `run: Optional[Run]`. One way would be to review each and every place where `Plan` is instantiated and find out whether `run` is even ever `None` - maybe it's always defined.

If it may be `None`, it'd be helpful to identify those situations, and we might split `my_run` into two attributes: one for regular use, which would apply `assert` internally and always return `Run`, and one for those special occasions, keeping the original `Optional[Run]` return type.

Or a subclass for those special cases, with a much narrower `my_run: None` (and changing the original to `Plan.my_run: Run`), to use them in different code flows. That would save us from having two different attributes I suggested above.

This requires a deeper knowledge of tmt internals than I have, and a dedicated review of code paths, and it's the reason why there's a "review all `assert`s" step in the annotations issue, so we could plan the next steps to reduce this very complexity.

File: tmt/base.py
Comment (happz) [reply tmt/base.py:121]: Right, I'll move it there.

File: tmt/base.py
Comment (happz) [reply tmt/base.py:496]: It's definitely not. I'm very unhappy about it. On the other hand, `export()` is already a bit too messy, with all those supported formats, that I didn't want to complicate it, even more, *in this patch*. This code belongs to `Test.export`, how about a `TODO` for now, and a more precise patch later?

File: tmt/base.py
Comment (happz) [reply tmt/base.py:83]: Fixed in b81d31a22671a9b3c691701bd1bcf16c84f9c4d1.

File: tmt/beakerlib.py
Comment (happz) [reply tmt/beakerlib.py]: Fixed in b81d31a22671a9b3c691701bd1bcf16c84f9c4d1.

File: tmt/base.py
Comment (happz) [reply tmt/base.py:276]: Fixed in b81d31a22671a9b3c691701bd1bcf16c84f9c4d1.

File: tmt/base.py
Comment (happz) [reply tmt/base.py]: Fixed in b81d31a22671a9b3c691701bd1bcf16c84f9c4d1.

File: tmt/cli.py
Comment (happz) [reply tmt/cli.py:1177]: I have a patch with a custom `Context` subclass ready, just in case :) It's 857 lines long, but oh boy, what a safety! :) 5 `cast()` calls were removed.

File: tmt/steps/discover/fmf.py
Comment (happz) [reply tmt/steps/discover/fmf.py:320]: One more thing I just ran into: adding annotations to fmf library could also help clear the situation. One of the repeating `assert`s is to narrow `fmf.Tree.root` to `str` - according to analysis, it's `Optional[str]` (it even starts as `self.root = None`).

File: tmt/cli.py
Comment (psss) [reply tmt/cli.py:715]: Yes, exactly.

File: tmt/base.py
Comment (psss) [reply tmt/base.py:509]: Understood, ack for a `TODO` for now.

File: tmt/steps/discover/fmf.py
Comment (psss) [reply tmt/steps/discover/fmf.py:299]: Hm, the `dist-git` part below checks for `fmf_root is None`, ok let's keep it as it is.

File: tmt/beakerlib.py
Comment (psss) [reply tmt/beakerlib.py:345]: Thanks for clarification.

File: tmt/cli.py
Comment (psss) [reply tmt/cli.py:1177]: Ok, let's rather keep the patch for another pull request, not to block this one ;-)

File: tmt/steps/discover/fmf.py
Comment (psss) [reply tmt/steps/discover/fmf.py:320]: Good to hear there are some options. Sounds like the extra subclass `PlanWithoutRun` with an explicit `my_run: None` could be a solution. Let's keep that for another PR as well.

File: tmt/cli.py
Comment (happz) [reply tmt/cli.py]: Fixed in 26ff555a1b94cd653b69baf38a5905d85ea9ecb0.

File: tmt/base.py
Comment (happz) [reply tmt/base.py:511]: Fixed in 26ff555a1b94cd653b69baf38a5905d85ea9ecb0.

File: tmt/cli.py
Comment (happz) [reply tmt/cli.py:1177]: Ok, submitted as an extra PR, https://github.com/teemtee/tmt/pull/1592.

File: tmt/base.py
Comment (psss) [inline tmt/base.py:1978]:
```suggestion
            tree = Tree(path)
```

File: tmt/base.py
Comment (psss) [inline tmt/base.py:2520]:
```suggestion
            # TODO: when generic, there would be no doubt about whether `clean_images`
```

**Review by psss — CHANGES_REQUESTED** [review]
A couple more comments.

File: tmt/base.py
Comment (happz) [reply tmt/base.py:476]: Because the `format_` parameter is of type `ExportFormat`, and that does include even formats the actual implementation does not support. There's no way to express this with just `ExportFormat`, we would need yet another type with just supported formats, and that seems like overkill.

In general, I consider the enum to be a temporary compromise. Since the export functionality might target different systems, I'd envision it as a more pluggable system, e.g. with plugins registering their "format" & a handler to perform the export. That would mitigate the need for output annotations depending on the output, and with a bit of abstraction, the enum & `@overload`s as well.

Comment (psss) [reply tmt/base.py:476]: I see, thanks much for the detailed clarification!

File: tmt/base.py
Comment (happz) [reply tmt/base.py]: Fixed in 3017568a8eed9792f95bd9528d2c0a22cedd84a6.

File: tmt/base.py
Comment (happz) [reply tmt/base.py:1635]: Fixed in 3017568a8eed9792f95bd9528d2c0a22cedd84a6.

File: tmt/base.py
Comment (happz) [reply tmt/base.py]: Fixed in 3017568a8eed9792f95bd9528d2c0a22cedd84a6.

File: tmt/base.py
Comment (happz) [reply tmt/base.py]: Fixed in 3017568a8eed9792f95bd9528d2c0a22cedd84a6.

File: tmt/base.py
Comment (happz) [reply tmt/base.py]: Fixed in 3017568a8eed9792f95bd9528d2c0a22cedd84a6.

**Review by psss — APPROVED** [review]
Thanks for addressing all comments! Looks good.

**Comment by psss** [comment]
Thanks for the `dict` format fix. Squashed. Should now be ready for merging if internal tests are green.

**Review by thrix — APPROVED** [review]
LGTM, oh wow! :rocket:

**Pull request merged**
